### PR TITLE
Command engine rework

### PR DESCRIPTION
**Pull request by samosaara**
This is not done yet, I'm creating the PR so I could get some input, and maybe testing from you guys as it progress, move the discussion from #79 to here. This is my TODO list so far.

- [x] Clear command
- [x] Random pokemon
- [x] Name filtering
- [x] Random pokemon from region
- [x] Random pokemon of type
- [x] Random pokemon of (configurable) lightness/darkness
- [x] Mixing filters
- [x] Wallpapers
- [x] Pokemon by ID
- [x] Testing
- [x] Documentation 
- [x] Slideshow

**IMPORTANT**: I'm currently using a bunch of python 3.6 features like type hints and `__init_subclass__`, if you guys don't support it, tell me before I use it too much.

### CODE REVIEW

**Comment by cclauss** [comment]
f'F-strings in Python 3.6 are really cool too'

Post an **issue** asking if users would be inconvenienced by a Py3.6 requirement.

Change .travis.yml from 3.5 to 3.6 or it will complain on every build.  PR 82 and 90 should help when they go in.

3.6 is not broadly deployed but this will be a good way to push more folks to adopt it.

I would recommend that you limit 3.6 dependencies to just one file (main.py?) so that it will be easy to backport if users push back.

File: filters/__init__.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+class DarkFilter(Filter):
+    def matches(self, pokemon: Pokemon, value):
+        return pokemon.get_dark_threshold() < value
+
+
+class TypeFilter(Filter):
+    def matches(self, pokemon: Pokemon, value):
+        return pokemon.get_pkmn_type() == value or\
+               pokemon.get_pkmn_type_secondary() == value
```
Comment (cclauss) [inline filters/__init__.py]: return value in (pokemon.get_pkmn_type(), pokemon.get_pkmn_type_secondary())

Looks really cool!

Comment (cclauss) [inline filters/__init__.py]: Perhaps pokemon should have store types as a two-tuple: (primary, secondary) with three related functions:
1. pokemon.get_pkmn_types()  # returns the two-tuple
2. pokemon.get_pkmn_type()   # returns two-tuple[0]
3. pokemon.get_pkmn_type_secondary()  # returns two-tuple[1]

Comment (samosaara) [inline filters/__init__.py]: I mean I think that barely makes a difference...

File: test_main.py [hunk 2]
```diff
@@ -3,97 +3,44 @@
...
+import filters
 from main import main
 from test_utils import region_dict
+import random
 
 db = Database()
 print(len(db))
 
 
-def test_no_args(capsys):
```
Comment (samosaara) [inline test_main.py]: This one is obsolete now because the new command engine does works with no arguments.

File: test_main.py [hunk 3]
```diff
@@ -3,97 +3,44 @@
...
-    main([__file__])
-    out, err = capsys.readouterr()
-    assert out.startswith("No command line arguments specified.")
-
-
-def test_three_args(capsys):
```
Comment (samosaara) [inline test_main.py]: Now with filter mixing the new command engine accepts way more then 3 arguments.

File: test_main.py [hunk 4]
```diff
@@ -3,97 +3,44 @@
...
-    main([__file__, 1, 2, 3])
-    out, err = capsys.readouterr()
-    assert out.startswith("Invalid number of arguments.")
-
-
-def test_two_letters(capsys):
```
Comment (samosaara) [inline test_main.py]: The 'search' feature is now built-in the name filter... No point...

File: test_main.py [hunk 5]
```diff
@@ -3,97 +3,44 @@
...
 
 
 def test_region_names(capsys):
-    main([__file__, 'regions'])
-    out, err = capsys.readouterr()
-    for region_name in region_dict:
-        assert region_name in out
-
-
-def test_help(capsys):
```
Comment (samosaara) [inline test_main.py]: I believe we can trust `argparse` here :sweat_smile:

File: test_main.py [hunk 6]
```diff
@@ -3,97 +3,44 @@
...
+    noExtras = filters.NonExtrasFilter()
     # matrix test of first pokemon name and last pokemon name from all regions
     for name, region_info in region_dict.items():
-        if name == 'extra':
-            continue
-        assert (region_info.first in out) == (name == region_name)
-        assert (region_info.last in out) == (name == region_name)
-
-
-def test_kanto(capsys):
```
Comment (samosaara) [inline test_main.py]: All these region specific testing have been merged into `region_test`

File: test_main.py [hunk 7]
```diff
@@ -3,97 +3,44 @@
...
-
-def test_all(capsys):
-    main([__file__, 'all'])
-    out, err = capsys.readouterr()
-    for region_info in region_dict.values():
-        assert (region_info.first or '') in out  # convert None --> ''
-        assert (region_info.last or '') in out   # convert None --> ''
-
-
-def test_question_mark(capsys):
```
Comment (samosaara) [inline test_main.py]: Not even exists anymore.

File: test_utils.py [hunk 8]
```diff
@@ -36,14 +33,6 @@ def expected_len(region_name):
     return region_info.end - region_info.start + 1
 
 
-def test_region_dict():
```
Comment (samosaara) [inline test_utils.py]: Do we really need to test the testing data? Like... ?!?!?

File: test_main.py (hunk 5 above)
Comment (samosaara) [inline test_main.py]: I mean bro, if we can't trust python's stdlib we need to start testing sums, boolean operations and built-in functions. Some assumptions must be made in order to test stuff.

I mean if we need to test that, we just might as well create test_python and see if basic arithmetic works

File: test_utils.py (hunk 8 above)
Comment (samosaara) [inline test_utils.py]: If we need to test the data we using to test our code there is no point in using the data in the first place. If the data is wrong the tests will fail how hard is that?

**Comment by cclauss** [comment]
Last time I looked, we execute **38 passed in 0.47 seconds** on every pull request.  You said that you really wanted automated testing and now you have it.  Please do not undermine it by deleting test cases that you did not write.  Having these automated tests in place will save us grief down the road -- I know that from long experience.  If you leave them in place, they might just catch corner cases that you are not thinking of today.

If test cases are in your way then please disable them in the way that I have mentioned above.  I do not want to get in your face, but I have been writing software for longer that you have been alive.  I have worked from some of the biggest, most important IT companies on the planet and my experience has been earned the hard way.  So please just trust me on this one.   If you want to understand, read the book http://www.obeythetestinggoat.com

**Comment by samosaara** [comment]
Sorry if I sounded a little bit too much pretentious there. Never really dig into testing before, since I didn't had much experience, some of these tests seemed very much `assert 1==1`.

First, some of the tests were based about a peculiar behavior the old command engine had. Since the behavior didn't exist anymore and that was a very specific thing I assumed it was OK to delete the tests. But I rolled those back.

But I just can't follow some of your logic, that '**we can't trust anyone**', I mean we are damned if we can't even trust python's stdlib. Like really what over tests we need to write? (1+1) == 2 that for me is a very paranoid approach to testing. That's why I don't agree we should test the help output... Well I would agree if we were formatting the output or something, but testing the default `argparse`'s output just seems pointless. The python stdlib has unit tests of their own. And about the '_Wikipedia test_' please, why on earth would we test the written numerical constants on the testing code?

But above all, please don't stand there and say that you are more experienced and therefore my criticism is void. Explain to me why I'm wrong.

**Comment by samosaara** [comment]
@LazoCoder This rework breaks all images of the readme... And I don't really have a Mac to redo them, what to do? (don't forget you have push permission to my fork). I'm finishing up the text but the images are all wrong now.

**Comment by LazoCoder** [comment]
Hey @samosaara, sorry that I have disappeared the past few weeks. I'm suffering from a mild RSI in my forearms so I have been minimizing my time spent on the computer to allow it to heal. Your contribution has not gone unnoticed. I want you to know that I have not forgotten about this. I *will* take a closer look at this. Also the readme looks fine, the images don't seem to be broken.

**Comment by samosaara** [comment]
OH silly me, i just assumed since i changed so much of the command engine the images would be all broken but stting the pokemon by name is still the same 

god speed, heal fast 😄

**Comment by samosaara** [comment]
So, when attempting to implement the slideshow feature I came across some doubts about how to implement it.
- First I took the liberty to _improve_ the current implementation by forking the process to the background and therefore not hanging the shell. It's great, better, but:
    - Even though I print the PID of the forked process to allow manual killing, I was not sure on how to implement a way to stop it through the script/program itself. 
    - My first thought was on to export the PID to a environment variable somehow, but that's not very 'n-shell compatible'
    - A file (even temporary in /tmp) is way too permanent to do that and I definitely don't want to keep a pid for too long and kill the wrong process later.
    - While writing about this I though about sockets or unix pipes... But that seems a little overkill... Even though i guess it would be the 'proper' way. And I don't know if Mac and GNU/linux are compatible on that front.
- I couldn't think of any way of being 'verbose' without invasive when it comes to slideshows.

**TL;DR**: How to properly kill the running slideshow daemon, and how to be verbose without annoying with the slideshows.

**Comment by LazoCoder** [comment]
Ok, I see what you mean about the screenshots. Do you think it would be a good idea to add a default functionality for a single parameter? For example if somebody does ```pokemon articuno``` then the ```-n``` is implied and it would be the same as doing ```pokemon -n articuno```?

Thats a good question about the slideshows. To be fair, I'm not sure what the best solution is, but you are right, its best to not be invasive (storing data to the disk) or to over complicate things (via pipes). Maybe just print the PID and let it be. The choice is yours.

**Comment by samosaara** [comment]
> Do you think it would be a good idea to add a default functionality for a single parameter? For example if somebody does pokemon articuno then the -n is implied and it would be the same as doing pokemon -n articuno?

Sure, actually a good idea, we can just try parsing the ID positional argument to a int and it fails convert it to a name filter instead.

About my last commit I ended up implementing the simplest pipe I could think of. I think the complexity wasn't half as bad as I was expecting. The good news though is that my implementation is somewhat more "_sturdy_"? And would be easy to expando to stuff like the status command and etc, because we now actually have a daemon thread listening to the pipe.

**Comment by samosaara** [comment]
I still, and now more then ever, need testing on the MacOS side of things. But by chance I noticed a idiosyncrasy on my name filtering that is, certain pokemon that have a shorter name, like abra, might match other unintended pokemon (e.g kadabra).
So to counter that, I'm thinking of instead of converting the ID to a name filter when not a number, It's going to only find (case insensitive) exact matches.

**Comment by samosaara** [comment]
I would like you guys to review it, especially @cclauss if he still has any issues with the testing I think now would be the time, or at least, review the changes other than the testing, so I can merge and you may create a more testing focused PR if you wish.
